OpenSim/Common/APDMDataReader.cpp

```diff
@@ -0,0 +1,243 @@
+        TimeSeriesTableVec3::RowVector
+            gyro_row_vector{ n_imus, SimTK::Vec3(SimTK::NaN) };
+        std::vector<std::string> nextRow = FileAdapter::getNextLine(in_stream, ",");
+        if (nextRow.empty()) {
+            done = true;
+            break;
+        }
+        // Cycle through the imus collating values
+        for (int imu_index = 0; imu_index < n_imus; ++imu_index) {
+            // parse gyro info from in_stream
+           if (foundLinearAccelerationData)
+                accel_row_vector[imu_index] = SimTK::Vec3(std::stod(nextRow[accIndex[imu_index]]),
+                    std::stod(nextRow[accIndex[imu_index] + 1]), std::stod(nextRow[accIndex[imu_index] + 2]));
+            if (foundMagneticHeadingData)
+                magneto_row_vector[imu_index] = SimTK::Vec3(std::stod(nextRow[magIndex[imu_index]]),
+                    std::stod(nextRow[magIndex[imu_index] + 1]), std::stod(nextRow[magIndex[imu_index] + 2]));
+            if (foundAngularVelocityData)
+                gyro_row_vector[imu_index] = SimTK::Vec3(std::stod(nextRow[gyroIndex[imu_index]]),
+                    std::stod(nextRow[gyroIndex[imu_index] + 1]), std::stod(nextRow[gyroIndex[imu_index] + 2]));
+            // Create Quaternion from values in file, assume order in file W, X, Y, Z
+            orientation_row_vector[imu_index] = 
+                SimTK::Quaternion(std::stod(nextRow[orientationsIndex[imu_index]]),
+                    std::stod(nextRow[orientationsIndex[imu_index] + 1]),
+                    std::stod(nextRow[orientationsIndex[imu_index] + 2]),
+                    std::stod(nextRow[orientationsIndex[imu_index] + 3]));
+        }
+        if (done) 
+            break;
+        // append to the tables
+        times[rowNumber] = time;
+        if (foundLinearAccelerationData) 
+            linearAccelerationData[rowNumber] =  accel_row_vector;
+        if (foundMagneticHeadingData) 
+            magneticHeadingData[rowNumber] = magneto_row_vector;
+        if (foundAngularVelocityData) 
+            angularVelocityData[rowNumber] = gyro_row_vector;
+        rotationsData[rowNumber] = orientation_row_vector;
+        // We could get some indication of time from file or generate time based on rate
+        // Here we use the latter mechanism.
+        time += timeIncrement;
+        rowNumber++;
+        if (std::remainder(rowNumber, last_size) == 0) {
+            // resize all Data/Matrices, double the size  while keeping data
+            int newSize = last_size*2;
+            times.resize(newSize);
+            // Repeat for Data matrices in use
+            if (foundLinearAccelerationData) linearAccelerationData.resizeKeep(newSize, n_imus);
+            if (foundMagneticHeadingData) magneticHeadingData.resizeKeep(newSize, n_imus);
+            if (foundAngularVelocityData) angularVelocityData.resizeKeep(newSize, n_imus);
+            rotationsData.resizeKeep(newSize, n_imus);
+            last_size = newSize;
```

Comment:
@chrisdembia The header has no indication of number of lines and these files are typically huge so we want to avoid scanning/reading them multiple times. This was a compromise that produced reasonable performance. While the data files in the example have size 1024, I tested with different initial_size to verify the code working and the test passed.